pkg/ccl/changefeedccl/changefeedbase/target.go

```diff
@@ -25,8 +25,102 @@ type Target struct {
 // the changefeed, possibly modified by WITH options.
 type StatementTimeName string
 
-// WatchedTables tracks the StatementTimeName for each table id.
-type WatchedTables map[descpb.ID]StatementTimeName
+type targetsByTable struct {
+	wholeTable   *Target
+	byFamilyName map[string]Target
+}
+
+func (tbt targetsByTable) add(t Target) targetsByTable {
+	if t.FamilyName == "" {
+		tbt.wholeTable = &t
+	} else {
+		if tbt.byFamilyName == nil {
+			tbt.byFamilyName = make(map[string]Target)
+		}
+		tbt.byFamilyName[t.FamilyName] = t
+	}
+	return tbt
+}
+
+func (tbt targetsByTable) each(f func(t Target) error) error {
+	if tbt.wholeTable != nil {
+		if err := f(*tbt.wholeTable); err != nil {
+			return err
+		}
+		for _, t := range tbt.byFamilyName {
+			if err := f(t); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
 
 // Targets is the complete list of target specifications for a changefeed.
-type Targets []Target
+// This is stored as a map of TableID -> Family Name -> Target in order
+// to support all current ways we need to iterate over it.
+type Targets struct {
+	Size uint
+	m    map[descpb.ID]targetsByTable
+}
+
+// Add adds a target to the list.
+func (ts *Targets) Add(t Target) {
+	if ts.m == nil {
+		ts.m = make(map[descpb.ID]targetsByTable)
+	}
+	ts.m[t.TableID] = ts.m[t.TableID].add(t)
+	ts.Size++
+}
+
+// EachTarget iterates over Targets.
+func (ts *Targets) EachTarget(f func(Target) error) error {
+	for _, l := range ts.m {
+		if err := l.each(f); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// EachTableID iterates over unique TableIDs referenced in Targets.
+func (ts *Targets) EachTableID(f func(descpb.ID) error) error {
```

Comment:
What we get from having this method is (free) uniqueness, there are two callers that needed to loop over all targets and extract unique table ids (because of `foo FAMILY f1, foo FAMILY f2` feeds).